### PR TITLE
store/tikv: validate range in CopClient.Send

### PR DESCRIPTION
**Pull request by winoros**
### What problem does this PR solve? 

Check range before sending requests.
To reduce the possibility of tikv panic.

### Check List 

Tests 

 - Unit test
 - Integration test (current tests all passed)

Side effects

 - Possible performance regression
 - Increased code complexity

---
This change is [Reviewable](https://reviewable.io/reviews/pingcap/tidb/8922)

### CODE REVIEW

**Comment by jackysp** [comment]
Could this be closed?

File: store/tikv/coprocessor.go [hunk 1]
```diff
@@ -109,6 +113,19 @@ func (c *CopClient) Send(ctx context.Context, req *kv.Request, vars *kv.Variable
 	return it
 }
 
+func (c *CopClient) validateRanges(ranges []kv.KeyRange) error {
+	for i := 1; i < len(ranges); i++ {
+		if bytes.Compare(ranges[i].StartKey, ranges[i-1].EndKey) < 0 {
+			startKeyLen := len(ranges[i].StartKey)
+			if startKeyLen+1 == len(ranges[i-1].EndKey) && bytes.Equal(ranges[i].StartKey, ranges[i-1].EndKey[:startKeyLen]) && ranges[i-1].EndKey[startKeyLen] == 0 {
```
Comment (winoros) [inline store/tikv/coprocessor.go]: @breeswish @zhangjinpeng1987  Please make sure this if branch is really a safe case for tikv.

Comment (winoros) [inline store/tikv/coprocessor.go]: @breeswish @zhangjinpeng1987  Please make sure this if branch is really a safe case for tikv.

**Comment by winoros** [comment]
```
➜  tikv git:(range-intersection-check) ✗ go test . -run=XXX -bench=BenchmarkCopClientValidateRanges -benchtime=5s

goos: darwin
goarch: amd64
pkg: github.com/pingcap/tidb/store/tikv
BenchmarkCopClientValidateRanges-8   	200000000	        31.2 ns/op
PASS
ok  	github.com/pingcap/tidb/store/tikv	9.418s
```

Local benchmark result on my mbp.

File: store/tikv/coprocessor.go (hunk 1 above)
Comment (breeswish) [reply store/tikv/coprocessor.go]: Does this actually violates start key < end key?

File: config/config.go [hunk 2]
```diff
@@ -61,6 +61,7 @@ type Config struct {
 	Socket           string          `toml:"socket" json:"socket"`
 	Lease            string          `toml:"lease" json:"lease"`
 	RunDDL           bool            `toml:"run-ddl" json:"run-ddl"`
+	RangeCheck       bool            `toml:"range-check" json:"range_check"`
```
Comment (lamxTyler) [inline config/config.go]: json:"range-check"`

File: store/tikv/coprocessor.go (hunk 1 above)
Comment (winoros) [reply store/tikv/coprocessor.go]: This is an exception for `NULL` value.

File: config/config.toml.example [hunk 3]
```diff
@@ -54,6 +54,9 @@ check-mb4-value-in-utf8 = true
 # treat-old-version-utf8-as-utf8mb4 use for upgrade compatibility. Set to true will treat old version table/column UTF8 charset as UTF8MB4.
 treat-old-version-utf8-as-utf8mb4 = true
 
+# range-check is used for enabling the validation of the range sending to tikv. Error will be thrown if the start key is not larger than the end key.
```
Comment (eurekaka) [inline config/config.toml.example]:
```suggestion
# range-check is used for enabling the validation of the range sending to TiKV. Error will be thrown if the start key of latter range is not larger than the end key of previous range.
```

File: store/tikv/coprocessor.go (hunk 1 above)
Comment (eurekaka) [reply store/tikv/coprocessor.go]: @winoros could you give an example of the `NULL` case for this if branch?

**Comment by zz-jason** [comment]
@winoros friendly ping, any update?

**Comment by zz-jason** [comment]
I'm going to close this PR due to outdated, fell free to reopen it if you are going to process it again in the future.

**Pull request closed without merging**